### PR TITLE
autoloader: Preload classes on plugin upgrade to avoid mid-upgrade fatals

### PR DESCRIPTION
**Pull request by anomiex**
#### Changes proposed in this Pull Request:

It's possible that the new version of the plugin will move some files
around that some later hook will try to load, causing a fatal as the
autoloader will still be looking in the old path. To avoid this, have
the Jetpack autoloader detect when an upgrade is happening and pre-load
all the classes in the plugin.

We do the same for plugin removal, for similar reasons.

Note this doesn't work for classes loaded via PSR-0 or PSR-4. But that
shouldn't matter for us since we always generate an optimized (classmap)
autoloader for releases of our plugins.

Also of note is that this won't work if you're manually installing a
specific version of the plugin as `wp plugin install jetpack --version=10.5`,
as wp-cli does that as an "install" rather than an "upgrade" operation.

#### Jetpack product discussion

p9dueE-4c4-p2

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Create a fresh site, e.g. with JN.
* Use Jetpack Beta to downgrade Jetpack to 10.4, and to install the release version of itself. Uninstall any other monorepo plugins. Also delete any `wp-content/plugins/*-dev`, just to be sure.
* Connect Jetpack (sync won't try to sync post-upgrade if you don't).
* Add the following mu-plugin:
  ```php
  <?php
  add_filter( 'pre_set_site_transient_update_plugins', function ( $transient ) {
  	if ( isset( $transient->response['jetpack/jetpack.php'] ) ) {
  		$transient->response['jetpack/jetpack.php']->new_version = '10.5';
  		$transient->response['jetpack/jetpack.php']->package = "https://downloads.wordpress.org/plugin/jetpack.10.5.zip";
  	}
  	return $transient;
  } );
  ```
* At this point the plugins page and `wp plugin upgrade` should both want to upgrade Jetpack to 10.5. Doing so should give a mid-upgrade fatal as described in p9dueE-4c4-p2.
* Re-install Jetpack 10.4.
* Use Jetpack Beta to check out the version of Jetpack Beta from this PR.
  * Optional: `wp shell <<<'spl_autoload_functions()[0];'` should indicate that it's using `Automattic\Jetpack\Autoloader\jp567fa3f555de8fd218dfdc1688bb97b5_betaⓥ3_1_1_alpha\PHP_Autoloader`.
* Now upgrading via the plugins page or `wp plugin upgrade` should succeed without a mid-upgrade fatal.

### CODE REVIEW

File: projects/packages/autoloader/src/class-autoloader.php [hunk 1]
```diff
@@ -77,6 +77,13 @@ public static function init( $container = null ) {
 		// Register a shutdown handler to clean up the autoloader.
 		$hook_manager->add_action( 'shutdown', new Shutdown_Handler( $plugins_handler, $cached_plugins, $was_included_by_autoloader ) );
 
+		// Register an upgrade handler to handler plugin upgrades.
```
Comment (fgiannar) [inline projects/packages/autoloader/src/class-autoloader.php:80]:
```suggestion
		// Register an upgrade handler to handle plugin upgrades.
```

**Review by fgiannar — CHANGES_REQUESTED** [review]
Hi Brad,

I tested using your instructions and everything works as expected as far as updating from JP 10.4 to JP 10.5 is concerned.
However, after successfully updating to 10.5 I removed the mu-plugin and tried updating to JP 10.6
This produced the following fatal:
```
PHP Fatal error:  require(): Failed opening required '(redacted)/public/wp-content/plugins/jetpack/vendor/automattic/jetpack-composer-plugin/src/class-plugin.php' (include_path='.:/opt/sp/php7.4/lib/php') in (redacted)/public/wp-content/plugins/jetpack-beta-dev/vendor/jetpack-autoloader/class-php-autoloader.php on line 90
```

**Comment by anomiex** [comment]
Good find! When we added the composer-plugin we didn't bother including it in the plugin bundle, but the autoloader still "knows" about its classes. Normally that doesn't matter since we don't actually load those classes, but this PR would try loading them.

Should be fixed now.

**Review by kbrown9 — COMMENTED** [review]
This approach looks good to me. As expected, commit [0981b47](https://github.com/Automattic/jetpack/pull/23039/commits/0981b47a99f37da9e8edeceeaf864bf562f3089d) did resolve the issue with attempting to load files that aren’t in the release in my tests.

**Review by fgiannar — DISMISSED** [review]
👍 Thanks for working on this! I repeated the testing instructions, plus attempting to update to 10.6 and 10.7 and succeeded without any errors.

**Comment by anomiex** [comment]
Needs re-approval after version bump

**Review by fgiannar — APPROVED** [review]

**Pull request merged**